File: DiscoTs/src/entities/Arquivo.ts
import { ItemArmazenamento } from './ItemArmazenamento';
import { Pasta } from './Pasta';
import fs from 'fs';

class Arquivo implements ItemArmazenamento {
  private nome: string;

  private tamanhoKB: number;

  private pastaPai: Pasta;

  private caminho: string;

  private tipo: string;

  constructor(nome: string, pastaPai: Pasta, caminho: string) {
    this.setNome(nome);
    this.setPastaPai(pastaPai);
    this.setCaminho(caminho);
    this.setTamanhoKB();
    this.setTipo();
  }

  public getNome(): string {
    return this.nome;
  }

  public setNome(nome: string): void {
    this.nome = nome;
  }

  private setTamanhoKB(): void {
    try {
      const infos = fs.statSync(this.caminho);

      if (infos.isFile()) {
        this.tamanhoKB = infos.size / 1024;
      } else {
        throw new Error('Caminho não encontrado');
      }
    } catch (e) {
      throw new Error('Caminho não encontrado');
    }
  }

  public getTamanhoKB(): Promise<number> {
    return Promise.resolve(this.tamanhoKB);
  }

  public getPastaPai(): Pasta {
    return this.pastaPai;
  }

  private setPastaPai(pastaPai: Pasta): void {
    this.pastaPai = pastaPai;
  }

  public getCaminho(): string {
    return this.caminho;
  }

  private setCaminho(caminho: string): void {
    this.caminho = caminho;
  }

  public async apagar(): Promise<void> {
    await this.pastaPai.remove(this);
  }

  public async add(item: ItemArmazenamento): Promise<void> {
    throw new Error('Arquivo não pode adicionar itens');
  }

  public async remove(item: ItemArmazenamento): Promise<void> {
    throw new Error('Arquivo não pode remover itens');
  }

  public async getItens(): Promise<ItemArmazenamento[]> {
    throw new Error('Arquivo não pode retornar itens');
  }

  public async getTipo(): Promise<string> {
    return this.tipo;
  }

  private setTipo(): void {
     
    const extensao = this.caminho.split('.');

    this.tipo = extensao[extensao.length - 1];

  }

}

export { Arquivo };

File: DiscoJava/DiscoVirtual/src/main/java/org/example/Arquivo.java
package org.example;

import java.io.File;
import java.util.List;

public class Arquivo implements ItemArmazenamento {

    private String nome;

    private String caminho;

    private Pasta pastaPai;
    private long tamanhoKB;

    public Arquivo(String nome, String caminho, Pasta pastaPai) {
        this.setNome(nome);
        this.setCaminho(caminho);
        this.setTamanhoKB();
        this.setPastaPai(pastaPai);
    }

    public String getNome() {
        return nome;
    }

    public void setNome(String nome) {
        this.nome = nome;
    }

    public long getTamanhoKB() {
        return tamanhoKB;
    }

    private void setTamanhoKB() {
        File file = new File(this.getCaminho());
        if( file.exists() ){
            this.tamanhoKB = file.length()/ 1024;
        }else{
            throw new IllegalArgumentException("Arquivo não encontrado");
        }
    }

    public Pasta getPastaPai() {
        return this.pastaPai;
    }

    private void setPastaPai(Pasta pastaPai) {
        this.pastaPai = pastaPai;
    }

    public String getCaminho() {
        return this.caminho;
    }

    private void setCaminho(String caminho) {
        this.caminho = caminho;
    }

    @Override
    public void apagar() {
        this.pastaPai.remove(this);
    }

    @Override
    public Boolean add(ItemArmazenamento item) {
        throw new UnsupportedOperationException("Arquivos não podem conter outros itens");
    }

    @Override
    public Boolean remove(ItemArmazenamento item) {
        throw new UnsupportedOperationException("Arquivos não podem conter outros itens");
    }

    @Override
    public List<ItemArmazenamento> getItens() {
        throw new UnsupportedOperationException("Arquivos não podem conter outros itens");
    }

    @Override
    public String toString() {
        return "Arquivo: " + this.getNome() + " - " + this.getTamanhoKB() + "KB";
    }

}
